### PR TITLE
(CDAP-1917) ClassLoader enhancement

### PR DESCRIPTION
**Pull request by chtyim**
This is to prepare for the ClassLoading needs for app template plugin class loading.

- Added a PackageFilterClassLoader
- Generalize MapReduceClassLoader to CombineClassLoader

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/lang/CombineClassLoader.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A {@link ClassLoader} that load classes from list of other {@link ClassLoader}s. Node that
```
Comment (nitinmotgi) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/CombineClassLoader.java]: Node => Note

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/CombineClassLoader.java]: Fixed.

File: cdap-common/src/main/java/co/cask/cdap/common/lang/PackageFilterClassLoader.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  private final Predicate<String> predicate;
+  private final ClassLoader bootstrapClassLoader;
+
+  /**
+   * Constructs a new instance that only allow class's package name passes the given {@link Predicate}.
+   */
+  public PackageFilterClassLoader(ClassLoader parent, Predicate<String> predicate) {
+    super(parent);
+    this.predicate = predicate;
+    this.bootstrapClassLoader = new URLClassLoader(new URL[0], null);
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/PackageFilterClassLoader.java]: I think this deserves a comment - why a URL[0]?

**Comment by anew** [comment]
LGTM except that I would like to see one comment.

File: cdap-common/src/main/java/co/cask/cdap/common/lang/PackageFilterClassLoader.java (hunk 2 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/PackageFilterClassLoader.java]: Added a comment.

**Comment by anew** [comment]
LGTM

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT1489-5

**Pull request merged**